crates/astria-sequencer/src/slinky/oracle/state_ext.rs

```diff
@@ -0,0 +1,289 @@
+        match bytes {
+            Some(bytes) => {
+                let currency_pair_state = serde_json::from_slice(&bytes)
+                    .context("failed to deserialize currency pair state")?;
+                Ok(Some(currency_pair_state))
+            }
+            None => Ok(None),
+        }
+    }
+
+    #[instrument(skip(self))]
+    async fn get_all_currency_pairs(&self) -> Result<Vec<CurrencyPair>> {
+        let prefix = format!("{CURRENCY_PAIR_STATE_PREFIX}/");
+        let mut currency_pairs: Vec<CurrencyPair> = Vec::new();
+
+        let mut stream = std::pin::pin!(self.prefix_keys(&prefix));
+        while let Some(Ok(key)) = stream.next().await {
+            let currency_pair = key
+                .strip_prefix(&prefix)
+                .context("failed to strip prefix from currency pair state key")?
+                .parse::<CurrencyPair>()
+                .context("failed to parse storage key suffix as currency pair")?;
+            currency_pairs.push(currency_pair);
+        }
+        Ok(currency_pairs)
+    }
+
+    #[instrument(skip(self))]
+    async fn get_next_currency_pair_id(&self) -> Result<u64> {
+        let Some(bytes) = self
+            .get_raw(NEXT_CURRENCY_PAIR_ID_KEY)
+            .await
+            .context("failed reading next currency pair id from state")?
+        else {
+            return Ok(0);
+        };
+        let Id(next_currency_pair_id) =
+            Id::try_from_slice(&bytes).context("invalid next currency pair id bytes")?;
+        Ok(next_currency_pair_id)
+    }
+
+    #[instrument(skip(self))]
+    async fn get_price_for_currency_pair(
+        &self,
+        currency_pair: &CurrencyPair,
+    ) -> Result<Option<QuotePrice>> {
+        let bytes = self
+            .get_raw(&currency_pair_price_storage_key(currency_pair))
+            .await
+            .context("failed to get price for currency pair from state")?;
+        match bytes {
+            Some(bytes) => {
+                let price =
+                    serde_json::from_slice(&bytes).context("failed to deserialize price")?;
+                Ok(Some(price))
+            }
+            None => Ok(None),
+        }
+    }
+}
+
+impl<T: StateRead + ?Sized> StateReadExt for T {}
+
+#[async_trait]
+pub(crate) trait StateWriteExt: StateWrite {
+    #[instrument(skip(self))]
+    fn put_currency_pair_id(&mut self, currency_pair: &CurrencyPair, id: u64) -> Result<()> {
+        let bytes = borsh::to_vec(&Id(id)).context("failed to serialize currency pair id")?;
+        self.put_raw(currency_pair_to_id_storage_key(currency_pair), bytes);
+        Ok(())
+    }
+
+    #[instrument(skip(self))]
+    fn put_currency_pair(&mut self, id: u64, currency_pair: CurrencyPair) -> Result<()> {
+        let bytes =
+            serde_json::to_vec(&currency_pair).context("failed to serialize currency pair")?;
+        self.put_raw(id_to_currency_pair_storage_key(id), bytes);
+        Ok(())
+    }
```

Comment:
These two methods might be better changed to free functions since they're not used outside of the trait?  Something like:
```rust
fn put_currency_pair_id<W: StateWrite + ?Sized>(
    state_writer: &mut W,
    currency_pair: &CurrencyPair,
    id: u64,
) -> Result<()> {
    ...
}
```